crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4640,10 +4650,413 @@ mod tests {
+        // the `x` definition in the `try` block must have been overridden after completion
+        // of the entire `try`/`except` block:
+        assert_public_ty(&db, "src/a.py", "x", "Literal[43, 44, 45]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_multiple_definitions_in_try_block_with_exhaustive_excepts(
+    ) -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 1
+                y = 2
+                z = 3
+            except NameError:
+                x = 11
+            except:
+                x = 12
+                y = 22
+            else:
+                x = 13
+                y = 23
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // TODO: What would be great in these assertions would be
+        // if we could assert the types of values *within* the `else` block.
+        // (Not possible with our current testing infrastructure, sadly.) --Alex
+
+        assert_public_ty(&db, "src/a.py", "x", "Literal[11, 12, 13]");
+
+        // `y` may be unbound if the `except NameError` branch is taken:
+        assert_public_ty(&db, "src/a.py", "y", "Unbound | Literal[2, 22, 23]");
+
+        // `z` may be unbound if the `except NameError` or the `except TypeError` branch is taken:
+        assert_public_ty(&db, "src/a.py", "z", "Unbound | Literal[3]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn nested_try_except_definitions() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                try:
+                    try:
+                        x = 1
+                    except TypeError:
+                        x = 2
+                    else:
+                        x = 3
+                except RuntimeError:
+                    x = 4
+                else:
+                    x = 5
+            except TypeError:
+                try:
+                    try:
+                        x = 6
+                    except RuntimeError:
+                        x = 7
+                    else:
+                        x = 8
+                except OSError:
+                    x = 9
+                else:
+                    x = 10
+            else:
+                try:
+                    try:
+                        x = 11
+                    except RuntimeError:
+                        x = 12
+                    else:
+                        x = 13
+                except OSError:
+                    x = 14
+                else:
+                    x = 15
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(
+            &db,
+            "src/a.py",
+            "x",
+            "Literal[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]",
```

Comment:
I think `x` can be `Unbound` here; it's never set in an exhaustive `except`, or in a `finally`.

I think this might be the issue I pointed out above, where we don't include the pre-try-block state as a pre-finally possibility.